### ISSUE TITLE
apiservice-kicker 100% CPU usage

### ISSUE THREAD

**Issue by matonb** [canonical/microk8s#2221]
Fresh install of microk8s (snap 1.21/edge) on PI4 (8GB), Ubuntu 21.04.

I noticed that apiservice-kickeris is taking 100% cpu on two nodes of a three node cluster.
The cluster isn't doing anything yet, I'm new to kubernetes and don't know where to look...

Any help appreciated.

[inspection-report-20210430_132345.tar.gz](https://github.com/ubuntu/microk8s/files/6405607/inspection-report-20210430_132345.tar.gz)

Saw the same issue on snap 1.21/stable

**Comment by ktsakalozos** [comment]
Hi @matonb, I cannot see anything wrong in your setup. I will need your help to figure out what may be causing this.

The apiservice-kicker [1] is a loop that runs every 3 seconds. In the loop there are a few tasks performed. Some of these tasks are "feature-flagged" via lock files found under `/var/snap/microk8s/current/var/lock`. Could you share the output of `ls -l  /var/snap/microk8s/current/var/lock`? We will need to run each tasks from the command line to see if some task gets stuck in an infinite loop.

[1] https://github.com/ubuntu/microk8s/blob/master/microk8s-resources/wrappers/apiservice-kicker

**Comment by matonb** [comment]
@ktsakalozos Here's the ouput:

```
ls -l /var/snap/microk8s/current/var/lock
total 0
-rw-r--r-- 1 root root     0 May  3 12:21 cni-loaded
-rw-rw---- 1 root microk8s 0 May  3 12:18 ha-cluster
-rw-rw---- 1 root microk8s 0 May  3 12:18 lite.lock
-rw-rw---- 1 root microk8s 0 May  3 12:18 no-etcd
-rw-rw---- 1 root microk8s 0 May  3 12:18 no-flanneld

```

**Comment by ktsakalozos** [comment]
Ok, we will need to go though the tasks in [1] one by one and see if any of them is causing this behavior.

Let's first disable the re-issuing of certificates on network changes, to do so edit `/var/snap/microk8s/current/arg/kube-apiserver` and append a line like `--advertise-address=127.0.0.1`, then restart microk8s with `microk8s.stop; microk8s.start`.

[1] https://github.com/ubuntu/microk8s/blob/master/microk8s-resources/wrappers/apiservice-kicker

**Comment by matonb** [comment]
Added to the top of the file:
```
head /var/snap/microk8s/current/args/kube-apiserver -n4
--advertise-address=127.0.0.1
--cert-dir=${SNAP_DATA}/certs
--service-cluster-ip-range=10.152.183.0/24
--authorization-mode=AlwaysAllow
```
microk8s restarted:
```
sudo su -
microk8s stop ; microk8s start
```

**Comment by ktsakalozos** [comment]
@matonb when you do a `sudo microk8s.stop` does the apiservice-kicker stop? I wonder if there is a leftover process.

**Comment by matonb** [comment]
@ktsakalozos
There's one `apiservice-kicker` process left running (100%) after I stop microk8s

**Comment by matonb** [comment]
On a node showing this behaviour (pid 131252 stuck at 100%):
```
ps -ef | grep kicker
root        1781       1  0 03:02 ?        00:00:40 /bin/bash /snap/microk8s/2175/apiservice-kicker
root      131252    1781 99 05:03 ?        04:11:53 /bin/bash /snap/microk8s/2175/apiservice-kicker
matonb    213917  211135  0 09:15 pts/0    00:00:00 grep --color=auto kicker
matonb@pi011:~$ sudo strace -f -p 1781
strace: Process 1781 attached
wait4(-1, ^Cstrace: Process 1781 detached
 <detached ...>

matonb@pi011:~$ sudo strace -f -p 131252
strace: Process 131252 attached

```

**Comment by matonb** [comment]
Can't terminate strace on stuck process, had to kill strace

```
sudo strace -f -p 131252
strace: Process 131252 attached
^C^C^C^CKilled
```

**Comment by matonb** [comment]
@ktsakalozos 
No suggestions?

I'll have to remove microk8s from all of the PI's I can't have it running at 100% CPU 24x7 / overheating before any services are added.

**Comment by ktsakalozos** [comment]
@matonb I cannot reproduce this issue. An I cannot figure out why this happens in your setup. Even if you restart the pi the stuck process comes up again with 100% CPU?

I understand the 100% CPU is not an acceptable setup.

**Comment by matonb** [comment]
It is very odd, it is happening on 2 of the 4 pi's in the cluster.
They're all Model 4 B 8GB, vanillia install of Ubuntu Server 21.04 64bit

Basically all that I've done to install is:
```
apt update
apt upgrade
snap install microk8s --classic --channel=1.21/edge

alias kubectl=/snap/bin/microk8s.kubectl
```
Picked a 'master' and joined the three others to it.

**Comment by nord73** [comment]
Same result here. PI4 8GB running Ubuntu 21.04 for a few weeks. I've even got two apiservice-kickers running after microk8s stop. Not in any practical use yet, but about to get started.

```
$ ps aux|grep apiservice|grep -v grep
root     2118288 99.4  0.0   9128   704 ?        R    May01 28377:11 /bin/bash /snap/microk8s/2175/apiservice-kicker
root     2870911 98.5  0.0   9260   752 ?        R    May18 3995:13 /bin/bash /snap/microk8s/2215/apiservice-kicker

$ uptime
 12:37:03 up 20 days, 20:17,  4 users,  load average: 2.00, 2.14, 2.78

$ date
Fri May 21 12:37:04 UTC 2021

$ sudo strace -f -p 2118288
strace: Process 2118288 attached

$ sudo strace -f -p 2118288
strace: Process 2118288 attached
^C^C^C^C^C^C^CKilled

$ sudo strace -f -p 2870911
strace: Process 2870911 attached
^C^C^C^C^CKilled

$ lsb_release -r
Release:        21.04

$ snap info microk8s|grep installed
installed:          v1.21.1             (2215) 168MB classic

```

**Comment by igorabiola** [comment]
this https://github.com/ubuntu/microk8s/issues/1775#issuecomment-734218451
seems to help. I'm still testing

**Comment by nord73** [comment]
> this https://github.com/ubuntu/microk8s/issues/1775#issuecomment-734218451
> 
> seems to help. I'm still testing

This did not work for me. Have since changed systemd's journal storage to volatile (based on recommendations in another thread and Ubuntu's docs) and watching for performance issues.

**Comment by nord73** [comment]
And the hanging apiservice-kicker process is back at 100%. I suppose I could try to replace the SD media with USB connected SSD's, but it kind of defeats the purpose of running a cluster on pis'.

**Comment by matonb** [comment]
SSD's won't fix the problem, that's what I'm running with on my PI's.

  I think the issue is network related, I've got a few other projects on an not had chance to look into this much.

**Comment by balchua** [comment]
@matonb and @nord73 there's also recently reported issue where users see this.
https://github.com/ubuntu/microk8s/issues/2280#issuecomment-851365650  based on their investigation moving the journald to volatile, improve disk storage and disabling metrics server seems to do the trick.

**Comment by nord73** [comment]
@balchua No metrics server here, and volatila storage for journald, so only storage performance left of those to consider. But as I mentioned earlier, taking built-in micro SD storage (even fast ones) out of the equation makes the idea of a pi cluster moot for most.

**Comment by nord73** [comment]
Still seeing this issue with v1.22.0.

**Comment by george-kirillov** [comment]
I also ran into this problem. PI4 8GB / Ubuntu 21.04 / usb ssd

**Comment by isilcala** [comment]
I also ran into this problem.  PI4 8GB / Ubuntu 21.04 / usb ssd

**Comment by xaerocom** [comment]
I've got the same issue Pi4 4GB / Ubuntu / Ubuntu 21.04 / sd card

**Comment by jontybrook** [comment]
I'm also seeing this - 4 x Pi 4 4GB running Ubuntu 21.04, microk8s 1.21. 

Here's a screenshot of glances on one of the nodes. 

![image](https://user-images.githubusercontent.com/542319/135339766-b0ecd915-9dd1-4604-abcd-087e011c14c6.png)

**Comment by fleshin** [comment]
I'm also seeing this - 4 x Pi 4 8GB. But now I see 2 processes stuck at 100%

**Comment by JoshuaDodds** [comment]
> @matonb I cannot reproduce this issue. An I cannot figure out why this happens in your setup. Even if you restart the pi the stuck process comes up again with 100% CPU?
> 
> I understand the 100% CPU is not an acceptable setup.

Hi, Please read this entire thread which is actually about this same issue.  There is no solve for it yet other than switching from ubuntu to debian and using faster storage but i dont think either of those "fixes" should be necessary.

**Comment by JoshuaDodds** [comment]
> @matonb I cannot reproduce this issue. An I cannot figure out why this happens in your setup. Even if you restart the pi the stuck process comes up again with 100% CPU?
> 
> I understand the 100% CPU is not an acceptable setup.

Hi @ktsakalozos, Please read this entire thread here (if you have not yet):
https://github.com/ubuntu/microk8s/issues/2280

Same issue and a lot of research has happened there as well. There is no solve for it yet other than switching from ubuntu to debian and using faster storage but i dont think either of those "fixes" should be necessary.

Also this might be related:
https://unix.stackexchange.com/questions/386776/unkillable-unstoppable-normal-user-process-how-to-forcedly-control-it

**Comment by bertrand1974** [comment]
I have the same issues as everyone else here. Running Ubuntu 21.10 on RPI 4/8GB, each booted off an SSD drive.

microk8s was installed using this command: `sudo snap install microk8s --classic --channel=1.21`. I have installed the metrics-server today, but prior to that, it wasn't running and the apiservice-kicker process was pegged at 100%. I will try installing debian and see if it fixes the issue or not.

**Comment by fleshin** [comment]
Is there an easy way to make that script modifiable in any way? I could try to debug on my end, but as it is part of the snap it is on a read only FS. Horrible idea!

**Comment by rmannibucau** [comment]
I got the same issue, apiservice-kicker using ~80% and systemd-udevd was also taking a high load of the CPU on a raspberry pi 400 - using ubuntu 21.10 server.
After having installed `linux-modules-extra-raspi`, CPU went down to something more expected/normal.

hope it helps a few

**Comment by JoshuaDodds** [comment]
I am wondering why nobody has bothered to look into this yest from canonical ?  Its still an issue....

**Comment by neoaggelos** [comment]
Hi @JoshuaDodds

This is in fact not ignored or abandoned by the MicroK8s team. Our testing concluded that it was an issue with the apiservice kicker using `ethtool` to check whether the vxlan.calico interface had a checksum offloading calculation enabled or disabled.

[This](https://github.com/ubuntu/microk8s/blob/1.23/microk8s-resources/wrappers/apiservice-kicker#L128-L140) has been removed [in newer versions](https://github.com/ubuntu/microk8s/blob/master/microk8s-resources/default-hooks/reconcile.d/10-cni-restart#L15-L20). Starting from MicroK8s 1.24, we are now using the configuration offered by Calico to properly disable the check instead of relying on the old behaviour, which would not cause the 100% CPU usage lockup.

In light of this, we have not seen the issue come up recently, but it is sometimes hard to follow up and we sometimes miss some updaes. Given that it has been a year since the issue was reported, I don't know if the original reporter can weigh in and check whether the are getting the same problem on 1.24+.

**Comment by JoshuaDodds** [comment]
Hi @neoaggelos , 

Much thanks for your quick response.  I have been following this issue since inception and somehow missed this update.  This is good news to hear that you believe that you have found the root cause of the issue.  I will assist when i have time by upgrading to 1.24 and re-enabling the api-kicker to test how it behaves.  I will report back with my findings.  

Thanks again for the quick response to my comment!
